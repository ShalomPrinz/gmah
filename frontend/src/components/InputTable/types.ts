interface TableColumn {
  id: number;
  label?: string;
  path: string;
}

type FormItem = { [key: string]: string };

type FormValues = {
  [key: string]: FormItem[];
};

export type { FormItem, FormValues, TableColumn };
