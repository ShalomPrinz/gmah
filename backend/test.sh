#!/bin/bash

if [[ $1 == "--coverage" ]]; then
    coverage run -m unittest discover tests
    coverage html
else
    python -m unittest discover tests
fi
