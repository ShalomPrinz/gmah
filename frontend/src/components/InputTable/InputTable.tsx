import {
  FieldErrors,
  FormProvider,
  useFieldArray,
  useForm,
} from "react-hook-form";
import { toast } from "react-toastify";
import { yupResolver } from "@hookform/resolvers/yup";

import IconComponent from "../Icon";
import TableBody from "./TableBody";
import TableHeader from "./TableHeader";
import type { FormItem, FormValues, TableColumn } from "./types";

import { formatFamilyKey } from "../../util";
import type { NonEmptyString } from "../../types";

interface InputTableProps {
  columns: TableColumn[];
  defaultItem: FormItem;
  initialValues: FormItem[];
  formName: NonEmptyString;
  onSubmit: (values: FormItem[]) => void;
  schema: any;
  text: {
    submit: string;
    /** When table has no values, noTable will be displayed */
    noTable: string;
  };
}

function InputTable({
  columns,
  defaultItem,
  initialValues,
  formName,
  onSubmit,
  schema,
  text: { submit, noTable },
}: InputTableProps) {
  const name = formName as string;
  const formMethods = useForm<FormValues>({
    defaultValues: {
      [name]: initialValues,
    },
    mode: "onBlur",
    resolver: yupResolver(schema),
  });

  // All form methods are required for Form Context Provider
  const {
    control,
    handleSubmit,
    getValues,
    formState: { isSubmitting },
  } = formMethods;
  const displayTable = getValues()?.[name]?.length > 0;
  const fieldArrayMethods = useFieldArray({
    name,
    control,
  });

  function onSubmitInit(e: React.FormEvent<HTMLFormElement>) {
    e.preventDefault();
    handleSubmit(
      (data) => onSubmit(data[name]),
      (errors) => handleErrorSubmission(errors)
    )();
  }

  function handleErrorSubmission(errors: FieldErrors<FormValues>) {
    if (!errors || !errors[name] || !Array.isArray(errors[name])) {
      toast.error("שגיאה לא צפויה");
      return;
    }

    // @ts-ignore typescript doesn't allow forEach here
    errors[name]?.forEach((familyErrors, index) => {
      familyErrors &&
        Object.entries(familyErrors).forEach(([field, error]) => {
          const key = formatFamilyKey(field);
          const message = error?.message || "שגיאה לא צפויה";
          toast.warning(
            `יש שגיאה ב${key} של משפחה מספר ${index + 1}: ${message}`,
            { toastId: `${index}:${field}` }
          );
        });
    });
  }

  return (
    <FormProvider {...formMethods}>
      <form onSubmit={onSubmitInit}>
        <table className="bg-white mx-auto">
          <TableHeader columns={columns} display={displayTable} />
          <TableBody
            columns={columns}
            fieldArrayMethods={fieldArrayMethods}
            formName={name}
          />
        </table>
        <button
          className="fs-3 bg-default rounded p-4 ms-5 mb-5 position-fixed bottom-0 start-0 button-hover"
          onClick={() => fieldArrayMethods.append(defaultItem)}
          type="button"
        >
          <span className="ps-3">הוסף שורה בטבלה</span>
          <IconComponent icon="addFamily" />
        </button>
        {displayTable && (
          <button
            className="fs-3 bg-default rounded m-4 py-3 px-4 button-border-focus"
            disabled={isSubmitting}
            type="submit"
          >
            {submit}
          </button>
        )}
        {!displayTable && <h1 className="mt-5 pt-4 fw-light">{noTable}</h1>}
      </form>
    </FormProvider>
  );
}

export default InputTable;
