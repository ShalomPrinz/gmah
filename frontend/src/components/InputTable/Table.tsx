import { FieldArray, Form, Formik } from "formik";
import type { FormikErrors } from "formik";
import { useEffect, useState } from "react";
import { toast } from "react-toastify";

import TableBody from "./TableBody";
import TableHeader from "./TableHeader";

import type { TableColumn } from "./types";
import IconComponent from "../Icon";

type TableRow = {
  [key: string]: string;
};

interface InputTableProps {
  columns: TableColumn[];
  /** Used for adding item to the table */
  emptyItem: TableRow;
  initialData: TableRow[];
  inputsName: string;
  onSubmit: (data: TableRow[]) => Promise<boolean>;
  /** Yup Object Schema with one key value pair: { [inputsName]: data } */
  schema: any;
}

const InputTable = ({
  columns,
  emptyItem,
  initialData,
  inputsName,
  onSubmit,
  schema,
}: InputTableProps) => {
  const { itemsCount, addItem, removeItem } = useItemsCount(
    initialData.length,
    emptyItem
  );
  const displayTable = itemsCount > 0;

  const toastUnknownSubmitError = () =>
    toast.error("קרתה תקלה בהוספת המשפחות", { toastId: "addFamiliesError" });

  const onSubmitResolved = (isSuccess: boolean) => {
    if (isSuccess) {
      toast.success(`הוספת בהצלחה ${itemsCount} משפחות חדשות לגמ"ח`);
    } else {
      toastUnknownSubmitError();
    }
  };

  function handleErrorSubmission(
    errors: FormikErrors<{ [x: string]: TableRow[] }>
  ) {
    if (!errors || !errors[inputsName] || !Array.isArray(errors[inputsName]))
      return;

    const errorsArray = errors[inputsName] as TableRow[];
    errorsArray.forEach((familyErrors, index) => {
      familyErrors &&
        Object.entries(familyErrors).forEach(([field, message]) =>
          toast.warning(
            `יש שגיאה ב${field} של משפחה מספר ${index + 1}: ${message}`,
            { toastId: `${index}:${field}` }
          )
        );
    });
  }

  return (
    <Formik
      enableReinitialize
      initialValues={{ [inputsName]: initialData }}
      validationSchema={schema}
      onSubmit={(values, { setSubmitting }) => {
        setSubmitting(true);
        onSubmit(values[inputsName])
          .then((isSuccess) => onSubmitResolved(isSuccess))
          .catch(() => toastUnknownSubmitError())
          .finally(() => setSubmitting(false));
      }}
    >
      {({ isSubmitting, errors }) => (
        <Form onSubmitCapture={() => handleErrorSubmission(errors)}>
          <FieldArray name={inputsName}>
            {({ push, remove }) => (
              <>
                <table className="bg-white">
                  <TableHeader columns={columns} display={displayTable} />
                  <TableBody
                    columns={columns}
                    inputsName={inputsName}
                    items={itemsCount}
                    removeFunc={(index: number) => removeItem(remove, index)}
                  />
                </table>
                <button
                  className="fs-3 bg-default rounded p-4 ms-5 mb-5 position-fixed bottom-0 start-0 button-hover"
                  onClick={() => addItem(push)}
                  type="button"
                >
                  <span className="ps-3">הוסף שורה בטבלה</span>
                  <IconComponent icon="addFamily" />
                </button>
              </>
            )}
          </FieldArray>
          {isSubmitting && (
            <span className="fs-4">הוספת המשפחות מתבצעת כעת, אנא המתן...</span>
          )}
          {displayTable && (
            <button
              className="bg-default p-4 rounded fs-3 button-border-focus m-5"
              disabled={isSubmitting}
              type="submit"
            >
              הוסף משפחות
            </button>
          )}
          {!displayTable && (
            <h1 className="mt-5 pt-4 fw-light">
              - כאן תוכל לערוך את המשפחות לאחר שתדביק את הטבלה -
            </h1>
          )}
        </Form>
      )}
    </Formik>
  );
};

function useItemsCount(initialCount: number, emptyItem: TableRow) {
  const [itemsCount, setItemsCount] = useState(initialCount);

  useEffect(() => setItemsCount(initialCount), [initialCount]);

  const addItem = (pushFunc: (obj: any) => void) => {
    pushFunc(emptyItem);
    setItemsCount((count) => count + 1);
  };

  const removeItem = (removeFunc: (index: number) => void, index: number) => {
    removeFunc(index);
    setItemsCount((count) => count - 1);
  };

  return { itemsCount, addItem, removeItem };
}

export default InputTable;
