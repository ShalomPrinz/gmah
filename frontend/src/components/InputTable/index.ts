import InputTable from "./Table";

export { InputTable };
