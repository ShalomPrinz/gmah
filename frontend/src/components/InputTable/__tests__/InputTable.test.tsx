import { describe, expect, it, vi } from "vitest";
import { array, object, string } from "yup";

import { render, screen, setupUser } from "../../../../test";

import InputTable from "../InputTable";

const emptyOnSubmit = () => Promise.resolve();

describe("InputTable", () => {
  describe("snapshot", () => {
    it("should render empty input table", () => {
      const { asFragment } = render(
        <InputTable
          columns={[]}
          defaultItem={{}}
          formName="empty"
          initialValues={[]}
          onSubmit={emptyOnSubmit}
          schema={object()}
          text={{
            noTable: "",
            submit: "",
          }}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render input table with one column", () => {
      // Table headers should not be displayed when table has no rows (body)
      const { asFragment } = render(
        <InputTable
          columns={[
            {
              id: 0,
              path: "name",
            },
          ]}
          defaultItem={{}}
          formName="oneColumn"
          initialValues={[]}
          onSubmit={emptyOnSubmit}
          schema={object()}
          text={{
            noTable: "",
            submit: "",
          }}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render input table with one column and one row", () => {
      const { asFragment } = render(
        <InputTable
          columns={[
            {
              id: 0,
              path: "name",
            },
          ]}
          defaultItem={{}}
          formName="oneColOneRow"
          initialValues={[
            {
              name: "Shalom Prinz",
            },
          ]}
          onSubmit={emptyOnSubmit}
          schema={object()}
          text={{
            noTable: "",
            submit: "",
          }}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render input table with two columns and three rows", () => {
      const { asFragment } = render(
        <InputTable
          columns={[
            {
              id: 0,
              path: "name",
            },
            {
              id: 1,
              path: "value",
            },
          ]}
          defaultItem={{}}
          formName="twoColsThreeRows"
          initialValues={[
            {
              name: "Shalom Prinz",
              value: "8",
            },
            {
              name: "Nitay Prinz",
              value: "21",
            },
            {
              name: "Someone",
              value: "17",
            },
          ]}
          onSubmit={emptyOnSubmit}
          schema={object()}
          text={{
            noTable: "",
            submit: "",
          }}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render input table with custom text when no table has no values", () => {
      const { asFragment } = render(
        <InputTable
          columns={[
            {
              id: 0,
              path: "name",
            },
          ]}
          defaultItem={{}}
          formName="customText"
          initialValues={[]}
          onSubmit={emptyOnSubmit}
          schema={object()}
          text={{
            noTable: "This text should show up when table has no values",
            submit: "should not show up",
          }}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render input table with custom submit text", () => {
      const { asFragment } = render(
        <InputTable
          columns={[
            {
              id: 0,
              path: "name",
            },
          ]}
          defaultItem={{}}
          formName="submitText"
          initialValues={[
            {
              name: "Anybody",
            },
          ]}
          onSubmit={emptyOnSubmit}
          schema={object()}
          text={{
            noTable: "should not show up",
            submit: "Submit Form",
          }}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("behavior", () => {
    it("should submit initial values", async () => {
      const user = setupUser();
      const submitText = "Submit Here";
      const onSubmit = vi.fn((values: object[]) => Promise.resolve());
      const initialValues = [{ id: "0", value: "Some data" }];
      render(
        <InputTable
          columns={[
            {
              id: 0,
              path: "value",
            },
          ]}
          defaultItem={{}}
          formName="testSubmit"
          initialValues={initialValues}
          onSubmit={onSubmit}
          schema={object()}
          text={{
            noTable: "",
            submit: submitText,
          }}
        />
      );

      const submitButton = screen.getByRole("button", { name: submitText });
      await user.click(submitButton);

      expect(onSubmit).toBeCalledWith(initialValues);
    });

    it("should submit types values over initial values", async () => {
      const user = setupUser();
      const submitText = "Submit Here";
      const onSubmit = vi.fn((values: object[]) => Promise.resolve());

      const valueId = "id";
      const initialValues = [{ id: valueId, value: "Some data" }];
      const typedValues = [{ id: valueId, value: "Shalom Prinz" }];

      render(
        <InputTable
          columns={[
            {
              id: 0,
              path: "value",
            },
          ]}
          defaultItem={{}}
          formName="testSubmit"
          initialValues={initialValues}
          onSubmit={onSubmit}
          schema={object()}
          text={{
            noTable: "",
            submit: submitText,
          }}
        />
      );

      const inputElement = screen.getByRole("textbox");
      await user.clear(inputElement);
      await user.type(inputElement, typedValues[0].value);

      const submitButton = screen.getByRole("button", { name: submitText });
      await user.click(submitButton);

      expect(onSubmit).toBeCalledWith(typedValues);
    });

    it("should not reset form values after successful submission", async () => {
      const user = setupUser();
      const submitText = "Submit Here";
      const initialValues = [{ id: "0", value: "Some data" }];
      render(
        <InputTable
          columns={[
            {
              id: 0,
              path: "value",
            },
          ]}
          defaultItem={{}}
          formName="testNoReset"
          initialValues={initialValues}
          onSubmit={emptyOnSubmit}
          schema={object()}
          text={{
            noTable: "",
            submit: submitText,
          }}
        />
      );

      const getInputValue = () =>
        screen.getByRole<HTMLInputElement>("textbox").value;

      expect(getInputValue()).toBe(initialValues[0].value);

      const submitButton = screen.getByRole("button", { name: submitText });
      await user.click(submitButton);

      expect(getInputValue()).toBe(initialValues[0].value);
    });

    it("should not submit if schema validation failed", async () => {
      const user = setupUser();
      const submitText = "Submit Here";
      const onSubmit = vi.fn((values: object[]) => Promise.resolve());

      const itemId = "0";
      const expectedString = "RightOne";
      const formName = "testSchemaFail";

      render(
        <InputTable
          columns={[
            {
              id: 0,
              path: "name",
            },
          ]}
          defaultItem={{}}
          formName={formName}
          initialValues={[{ id: itemId, name: "" }]}
          onSubmit={onSubmit}
          schema={object({
            [formName]: array().of(
              object({
                name: string().matches(new RegExp(`^${expectedString}$`)),
              })
            ),
          })}
          text={{
            noTable: "",
            submit: submitText,
          }}
        />
      );

      const inputElement = screen.getByRole("textbox");
      await user.type(inputElement, "incorrect");

      const submitButton = screen.getByRole("button", { name: submitText });
      await user.click(submitButton);

      expect(onSubmit).not.toBeCalled();

      await user.clear(inputElement);
      await user.type(inputElement, expectedString);
      await user.click(submitButton);

      expect(onSubmit).toBeCalledWith([{ id: itemId, name: expectedString }]);
    });

    it("should display error in the DOM if schema validation failed", async () => {
      const user = setupUser();
      const submitText = "Submit Here";
      const onSubmit = vi.fn((values: object[]) => Promise.resolve());

      const itemId = "0";
      const expectedString = "RightOne";
      const formName = "testSchemaFail";

      const { asFragment } = render(
        <InputTable
          columns={[
            {
              id: 0,
              path: "name",
            },
          ]}
          defaultItem={{}}
          formName={formName}
          initialValues={[{ id: itemId, name: "" }]}
          onSubmit={onSubmit}
          schema={object({
            [formName]: array().of(
              object({
                name: string().matches(new RegExp(`^${expectedString}$`)),
              })
            ),
          })}
          text={{
            noTable: "",
            submit: submitText,
          }}
        />
      );

      const inputElement = screen.getByRole("textbox");
      await user.type(inputElement, "incorrect");
      await user.tab(); // Unfocus element should cause validation

      const schemaFailSnapshot = asFragment();

      await user.clear(inputElement);
      await user.type(inputElement, expectedString);
      await user.tab(); // Unfocus element should cause validation

      expect(asFragment()).not.toEqual(schemaFailSnapshot);
    });

    it("should add row to the table", async () => {
      const user = setupUser();
      render(
        <InputTable
          columns={[
            {
              id: 0,
              path: "value",
            },
          ]}
          defaultItem={{
            value: "",
          }}
          formName="testAddRow"
          initialValues={[]}
          onSubmit={emptyOnSubmit}
          schema={object()}
          text={{
            noTable: "",
            submit: "",
          }}
        />
      );

      const getAllInputs = () => screen.queryAllByRole("textbox");

      expect(getAllInputs()).toHaveLength(0);

      const addRowButton = screen.getByRole("button", {
        name: "הוסף שורה בטבלה",
      });
      await user.click(addRowButton);

      expect(getAllInputs()).toHaveLength(1);
    });

    it("should add row to the table with initial data", async () => {
      const user = setupUser();
      render(
        <InputTable
          columns={[
            {
              id: 0,
              path: "value",
            },
          ]}
          defaultItem={{
            value: "",
          }}
          formName="testAppendRow"
          initialValues={[
            {
              id: "0",
              value: "First Row",
            },
          ]}
          onSubmit={emptyOnSubmit}
          schema={object()}
          text={{
            noTable: "",
            submit: "",
          }}
        />
      );

      const getAllInputs = () => screen.queryAllByRole("textbox");

      expect(getAllInputs()).toHaveLength(1);

      const addRowButton = screen.getByRole("button", {
        name: "הוסף שורה בטבלה",
      });
      await user.click(addRowButton);

      expect(getAllInputs()).toHaveLength(2);
    });
  });
});
