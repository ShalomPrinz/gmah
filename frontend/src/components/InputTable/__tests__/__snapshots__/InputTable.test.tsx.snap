// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputTable > snapshot > should render empty input table 1`] = `
<DocumentFragment>
  <form>
    <table
      class="bg-white mx-auto"
    >
      <thead>
        <tr />
      </thead>
      <tbody />
    </table>
    <button
      class="fs-3 bg-default rounded p-4 ms-5 mb-5 position-fixed bottom-0 start-0 button-hover"
      type="button"
    >
      <span
        class="ps-3"
      >
        הוסף שורה בטבלה
      </span>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-user-plus "
        color="white"
        data-icon="user-plus"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 640 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 128a128 128 0 1 1 256 0A128 128 0 1 1 96 128zM0 482.3C0 383.8 79.8 304 178.3 304h91.4C368.2 304 448 383.8 448 482.3c0 16.4-13.3 29.7-29.7 29.7H29.7C13.3 512 0 498.7 0 482.3zM504 312V248H440c-13.3 0-24-10.7-24-24s10.7-24 24-24h64V136c0-13.3 10.7-24 24-24s24 10.7 24 24v64h64c13.3 0 24 10.7 24 24s-10.7 24-24 24H552v64c0 13.3-10.7 24-24 24s-24-10.7-24-24z"
          fill="currentColor"
        />
      </svg>
    </button>
    <h1
      class="mt-5 pt-4 fw-light"
    />
  </form>
</DocumentFragment>
`;

exports[`InputTable > snapshot > should render input table with custom submit text 1`] = `
<DocumentFragment>
  <form>
    <table
      class="bg-white mx-auto"
    >
      <thead>
        <tr>
          <th />
          <th
            class="fs-5 p-3"
          >
            name
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="fs-4 fw-bold ps-3"
          >
            1
          </td>
          <td
            class="align-middle"
          >
            <input
              class="form-text-input p-1 m-1 text-center"
              name="submitText.0.name"
              style="max-width: 130px;"
              title="name"
              type="text"
            />
          </td>
          <td>
            <button
              class="me-2 bg-white text-danger rounded fs-5 border border-3 border-danger button-hover"
              type="button"
            >
              <span
                class="ps-2"
              >
                הסר
              </span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-circle-minus "
                color="red"
                data-icon="circle-minus"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 512c141.4 0 256-114.6 256-256S397.4 0 256 0S0 114.6 0 256S114.6 512 256 512zM184 232H328c13.3 0 24 10.7 24 24s-10.7 24-24 24H184c-13.3 0-24-10.7-24-24s10.7-24 24-24z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </td>
        </tr>
      </tbody>
    </table>
    <button
      class="fs-3 bg-default rounded p-4 ms-5 mb-5 position-fixed bottom-0 start-0 button-hover"
      type="button"
    >
      <span
        class="ps-3"
      >
        הוסף שורה בטבלה
      </span>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-user-plus "
        color="white"
        data-icon="user-plus"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 640 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 128a128 128 0 1 1 256 0A128 128 0 1 1 96 128zM0 482.3C0 383.8 79.8 304 178.3 304h91.4C368.2 304 448 383.8 448 482.3c0 16.4-13.3 29.7-29.7 29.7H29.7C13.3 512 0 498.7 0 482.3zM504 312V248H440c-13.3 0-24-10.7-24-24s10.7-24 24-24h64V136c0-13.3 10.7-24 24-24s24 10.7 24 24v64h64c13.3 0 24 10.7 24 24s-10.7 24-24 24H552v64c0 13.3-10.7 24-24 24s-24-10.7-24-24z"
          fill="currentColor"
        />
      </svg>
    </button>
    <button
      class="fs-3 bg-default rounded m-4 py-3 px-4 button-border-focus"
      type="submit"
    >
      Submit Form
    </button>
  </form>
</DocumentFragment>
`;

exports[`InputTable > snapshot > should render input table with custom text when no table has no values 1`] = `
<DocumentFragment>
  <form>
    <table
      class="bg-white mx-auto"
    >
      <thead>
        <tr />
      </thead>
      <tbody />
    </table>
    <button
      class="fs-3 bg-default rounded p-4 ms-5 mb-5 position-fixed bottom-0 start-0 button-hover"
      type="button"
    >
      <span
        class="ps-3"
      >
        הוסף שורה בטבלה
      </span>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-user-plus "
        color="white"
        data-icon="user-plus"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 640 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 128a128 128 0 1 1 256 0A128 128 0 1 1 96 128zM0 482.3C0 383.8 79.8 304 178.3 304h91.4C368.2 304 448 383.8 448 482.3c0 16.4-13.3 29.7-29.7 29.7H29.7C13.3 512 0 498.7 0 482.3zM504 312V248H440c-13.3 0-24-10.7-24-24s10.7-24 24-24h64V136c0-13.3 10.7-24 24-24s24 10.7 24 24v64h64c13.3 0 24 10.7 24 24s-10.7 24-24 24H552v64c0 13.3-10.7 24-24 24s-24-10.7-24-24z"
          fill="currentColor"
        />
      </svg>
    </button>
    <h1
      class="mt-5 pt-4 fw-light"
    >
      This text should show up when table has no values
    </h1>
  </form>
</DocumentFragment>
`;

exports[`InputTable > snapshot > should render input table with one column 1`] = `
<DocumentFragment>
  <form>
    <table
      class="bg-white mx-auto"
    >
      <thead>
        <tr />
      </thead>
      <tbody />
    </table>
    <button
      class="fs-3 bg-default rounded p-4 ms-5 mb-5 position-fixed bottom-0 start-0 button-hover"
      type="button"
    >
      <span
        class="ps-3"
      >
        הוסף שורה בטבלה
      </span>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-user-plus "
        color="white"
        data-icon="user-plus"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 640 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 128a128 128 0 1 1 256 0A128 128 0 1 1 96 128zM0 482.3C0 383.8 79.8 304 178.3 304h91.4C368.2 304 448 383.8 448 482.3c0 16.4-13.3 29.7-29.7 29.7H29.7C13.3 512 0 498.7 0 482.3zM504 312V248H440c-13.3 0-24-10.7-24-24s10.7-24 24-24h64V136c0-13.3 10.7-24 24-24s24 10.7 24 24v64h64c13.3 0 24 10.7 24 24s-10.7 24-24 24H552v64c0 13.3-10.7 24-24 24s-24-10.7-24-24z"
          fill="currentColor"
        />
      </svg>
    </button>
    <h1
      class="mt-5 pt-4 fw-light"
    />
  </form>
</DocumentFragment>
`;

exports[`InputTable > snapshot > should render input table with one column and one row 1`] = `
<DocumentFragment>
  <form>
    <table
      class="bg-white mx-auto"
    >
      <thead>
        <tr>
          <th />
          <th
            class="fs-5 p-3"
          >
            name
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="fs-4 fw-bold ps-3"
          >
            1
          </td>
          <td
            class="align-middle"
          >
            <input
              class="form-text-input p-1 m-1 text-center"
              name="oneColOneRow.0.name"
              style="max-width: 130px;"
              title="name"
              type="text"
            />
          </td>
          <td>
            <button
              class="me-2 bg-white text-danger rounded fs-5 border border-3 border-danger button-hover"
              type="button"
            >
              <span
                class="ps-2"
              >
                הסר
              </span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-circle-minus "
                color="red"
                data-icon="circle-minus"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 512c141.4 0 256-114.6 256-256S397.4 0 256 0S0 114.6 0 256S114.6 512 256 512zM184 232H328c13.3 0 24 10.7 24 24s-10.7 24-24 24H184c-13.3 0-24-10.7-24-24s10.7-24 24-24z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </td>
        </tr>
      </tbody>
    </table>
    <button
      class="fs-3 bg-default rounded p-4 ms-5 mb-5 position-fixed bottom-0 start-0 button-hover"
      type="button"
    >
      <span
        class="ps-3"
      >
        הוסף שורה בטבלה
      </span>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-user-plus "
        color="white"
        data-icon="user-plus"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 640 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 128a128 128 0 1 1 256 0A128 128 0 1 1 96 128zM0 482.3C0 383.8 79.8 304 178.3 304h91.4C368.2 304 448 383.8 448 482.3c0 16.4-13.3 29.7-29.7 29.7H29.7C13.3 512 0 498.7 0 482.3zM504 312V248H440c-13.3 0-24-10.7-24-24s10.7-24 24-24h64V136c0-13.3 10.7-24 24-24s24 10.7 24 24v64h64c13.3 0 24 10.7 24 24s-10.7 24-24 24H552v64c0 13.3-10.7 24-24 24s-24-10.7-24-24z"
          fill="currentColor"
        />
      </svg>
    </button>
    <button
      class="fs-3 bg-default rounded m-4 py-3 px-4 button-border-focus"
      type="submit"
    />
  </form>
</DocumentFragment>
`;

exports[`InputTable > snapshot > should render input table with two columns and three rows 1`] = `
<DocumentFragment>
  <form>
    <table
      class="bg-white mx-auto"
    >
      <thead>
        <tr>
          <th />
          <th
            class="fs-5 p-3"
          >
            name
          </th>
          <th
            class="fs-5 p-3"
          >
            value
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="fs-4 fw-bold ps-3"
          >
            1
          </td>
          <td
            class="align-middle"
          >
            <input
              class="form-text-input p-1 m-1 text-center"
              name="twoColsThreeRows.0.name"
              style="max-width: 130px;"
              title="name"
              type="text"
            />
          </td>
          <td
            class="align-middle"
          >
            <input
              class="form-text-input p-1 m-1 text-center"
              name="twoColsThreeRows.0.value"
              style="max-width: 130px;"
              title="value"
              type="text"
            />
          </td>
          <td>
            <button
              class="me-2 bg-white text-danger rounded fs-5 border border-3 border-danger button-hover"
              type="button"
            >
              <span
                class="ps-2"
              >
                הסר
              </span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-circle-minus "
                color="red"
                data-icon="circle-minus"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 512c141.4 0 256-114.6 256-256S397.4 0 256 0S0 114.6 0 256S114.6 512 256 512zM184 232H328c13.3 0 24 10.7 24 24s-10.7 24-24 24H184c-13.3 0-24-10.7-24-24s10.7-24 24-24z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </td>
        </tr>
        <tr>
          <td
            class="fs-4 fw-bold ps-3"
          >
            2
          </td>
          <td
            class="align-middle"
          >
            <input
              class="form-text-input p-1 m-1 text-center"
              name="twoColsThreeRows.1.name"
              style="max-width: 130px;"
              title="name"
              type="text"
            />
          </td>
          <td
            class="align-middle"
          >
            <input
              class="form-text-input p-1 m-1 text-center"
              name="twoColsThreeRows.1.value"
              style="max-width: 130px;"
              title="value"
              type="text"
            />
          </td>
          <td>
            <button
              class="me-2 bg-white text-danger rounded fs-5 border border-3 border-danger button-hover"
              type="button"
            >
              <span
                class="ps-2"
              >
                הסר
              </span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-circle-minus "
                color="red"
                data-icon="circle-minus"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 512c141.4 0 256-114.6 256-256S397.4 0 256 0S0 114.6 0 256S114.6 512 256 512zM184 232H328c13.3 0 24 10.7 24 24s-10.7 24-24 24H184c-13.3 0-24-10.7-24-24s10.7-24 24-24z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </td>
        </tr>
        <tr>
          <td
            class="fs-4 fw-bold ps-3"
          >
            3
          </td>
          <td
            class="align-middle"
          >
            <input
              class="form-text-input p-1 m-1 text-center"
              name="twoColsThreeRows.2.name"
              style="max-width: 130px;"
              title="name"
              type="text"
            />
          </td>
          <td
            class="align-middle"
          >
            <input
              class="form-text-input p-1 m-1 text-center"
              name="twoColsThreeRows.2.value"
              style="max-width: 130px;"
              title="value"
              type="text"
            />
          </td>
          <td>
            <button
              class="me-2 bg-white text-danger rounded fs-5 border border-3 border-danger button-hover"
              type="button"
            >
              <span
                class="ps-2"
              >
                הסר
              </span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-circle-minus "
                color="red"
                data-icon="circle-minus"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 512c141.4 0 256-114.6 256-256S397.4 0 256 0S0 114.6 0 256S114.6 512 256 512zM184 232H328c13.3 0 24 10.7 24 24s-10.7 24-24 24H184c-13.3 0-24-10.7-24-24s10.7-24 24-24z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </td>
        </tr>
      </tbody>
    </table>
    <button
      class="fs-3 bg-default rounded p-4 ms-5 mb-5 position-fixed bottom-0 start-0 button-hover"
      type="button"
    >
      <span
        class="ps-3"
      >
        הוסף שורה בטבלה
      </span>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-user-plus "
        color="white"
        data-icon="user-plus"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 640 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 128a128 128 0 1 1 256 0A128 128 0 1 1 96 128zM0 482.3C0 383.8 79.8 304 178.3 304h91.4C368.2 304 448 383.8 448 482.3c0 16.4-13.3 29.7-29.7 29.7H29.7C13.3 512 0 498.7 0 482.3zM504 312V248H440c-13.3 0-24-10.7-24-24s10.7-24 24-24h64V136c0-13.3 10.7-24 24-24s24 10.7 24 24v64h64c13.3 0 24 10.7 24 24s-10.7 24-24 24H552v64c0 13.3-10.7 24-24 24s-24-10.7-24-24z"
          fill="currentColor"
        />
      </svg>
    </button>
    <button
      class="fs-3 bg-default rounded m-4 py-3 px-4 button-border-focus"
      type="submit"
    />
  </form>
</DocumentFragment>
`;
